New k8s_facts module

willthames opened the pull request:
##### SUMMARY
Provide a very simple k8s_facts that mostly just reuses existing
code.

Deprecate k8s modules with kind=*List in favour of k8s_facts

Provide label selector functionality through k8s_facts

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
k8s_facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (devel 02135d5278) last updated 2018/06/05 22:26:48 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION

We don't necessarily need to deprecate `kind=*List` behaviour, but as it was broken previously (see #41088 on which this PR is based) and the k8s modules are very new, I don't see much harm in moving to a more familiar mechanism that meets Ansible best practices.

mattclay commented:
bot_status

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -60,24 +61,31 @@ def __init__(self, *args, **kwargs):
             search_kind = kind
             if kind.lower().endswith('list'):
                 search_kind = kind[:-4]
+                self.deprecate("Use the k8s_facts module with kind={0} rather than {1} with "
```
How would this deprecation work if we implement kind=*List the same way other clients do (by processing each element in the list and doing a create/get/delete/update)?

willthames replied:
Do you think that's a thing that's likely to happen? Right now, the List mechanism is only documented to return facts, and causes other issues when used with patching (hence my original bug report in #41088). If there's a genuine plan to repurpose `List` for patching purposes, we can not deprecate, but I'd favour using k8s_facts to get the list, and then applying k8s with a loop.

fabianvf replied:
yeah I think *List will definitely get implemented, I haven't found the time yet but the goal is to match existing clients to maximize portability.

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -91,24 +99,29 @@ def execute_module(self):
 
         try:
-            existing = resource.get(name=name, namespace=namespace)
+            existing = resource.get(name=name, namespace=namespace, label_selector=labels)
```
If we want to support selectors we should probably just let the user pass them in, and the client supports fieldSelectors as well as labelSelectors

fabianvf commented on lib/ansible/module_utils/k8s/common.py:
```diff
@@ -76,6 +76,10 @@
         'default': 'v1',
         'aliases': ['api', 'version'],
     },
+    'labels': {
```
I don't think `labels` belongs in the common arg spec, it doesn't mean anything outside of `k8s_facts`

fabianvf reviewed (changes requested):
It would make sense to make `k8s_facts` more divergent from `k8s_raw`, I don't think they have that much overlap. Most of the shared logic needed should be in `module_utils/k8s/common.py`. We could define a smaller interface on `k8s_facts` than we'd need for `k8s`, likely consisting of `api_version`, `group`, `kind`, `namespace`, `name`, `label_selector`, `field_selector`.

It should be able to use a lot of the same logic as the lookup plugin as well.

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -91,24 +99,29 @@ def execute_module(self):
         force = self.params.pop('force', False)
         name = definition.get('metadata', {}).get('name')
         namespace = definition.get('metadata', {}).get('namespace')
+        labels = ','.join(['%s=%s' % (k, v) for (k, v) in definition.get('metadata', {}).get('labels', {}).items()])
```
Kubernetes also supports `!=` and set based label selectors. It would probably be easier to have the user specify a selector and pass it through

willthames commented:
@fabianvf thanks for all your feedback. I've completely revised the module now, it should be more in line with your thoughts.

willthames commented:
ready_for_review

maxamillion commented on lib/ansible/modules/clustering/k8s/k8s_facts.py:
```diff
@@ -0,0 +1,171 @@
+
+    @property
+    def argspec(self):
+        args = copy.deepcopy(AUTH_ARG_SPEC)
```
Is there any concern of potentially leaking any sensitive information here if someone bumps verbosity for ansible output? Anything need to be marked `no_log`?

willthames replied:
That's already covered in `AUTH_ARG_SPEC`: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/k8s/common.py#L87-L92

maxamillion replied:
I completely missed that, apologies

maxamillion commented:
shipit

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -60,24 +60,30 @@ def __init__(self, *args, **kwargs):
 
     def execute_module(self):
         changed = False
+        facts = False
```
Do you still need any of the changes to this file? Looks like they're no longer used.

willthames replied:
It, or some alternative implementation, is still needed, as `List` still gets stripped off `kind` before the call to `perform_action` (which means that the bug where passing `List` causes a patch to be attempted still occurs because it doesn't know not to attempt a patch). The alternative is not to strip `List` off `kind` - really you're the best person to answer that as the code is authored by you.

I can completely strip out these changes and leave the original bug, now that I can use `k8s_facts` rather than `k8s` with `kind`=`*List` - depends what you want to do about the bug.

willthames replied:
I've removed the deprecation though.

willthames commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -101,8 +103,8 @@ def perform_action(self, resource, definition):
 
         self.remove_aliases()
 
-        if definition['kind'].endswith('list'):
```
@fabianvf - this would never ever be true, so this is the bit that still needs a fix if the `facts` parameter is removed from `perform_action`

fabianvf replied:
Good point. Wonder if we should just make the *List types fail out until we've properly implemented them.

fabianvf replied:
Regardless I think we can address that in a separate PR, I think for this PR we can add the `k8s_facts` module and leave the behavior of the other modules unchanged.

willthames replied:
All changes to raw.py have now been reverted.

Commit:
New k8s_facts module

Strip out the facts parts of the k8s module into a new simpler
module that deals with querying for facts, also allowing
field and label selectors

maxamillion commented:
rekicking CI, failure unrelated

willthames commented:
!needs_revision
